File: ocaml_revised/GCBench.ml
(*
Sun -
Two problems:
1. Tried to write up the code as similar as possible
   from GCBench.sml, but for some reason the building-up trees
   takes too much time
2. printing memory available (explained below )

*)


type tree = Empty
          | Node of node
            and node = { left: tree ref; right: tree ref; i: int; j: int}

let make_empty_node () =
  Node { left = ref Empty; right= ref Empty; i = 0 ; j = 0 }

let make_node (l, r) =
  Node { left  = ref l; right = ref r; i = 0 ; j = 0 }


(* Source:
http://caml.inria.fr/pub/docs/manual-ocaml/libref/Gc.html   *)
(* Sun: I wasn't able to find a method to represent
 "total memory available" by OCaml (Doesn't seem to have one in OCaml)
 Instead I found a way to represent  "total memory allocated"    *)
let printDiagnostics () = (
  let lTotalMemory =  8. *. ( (Gc.stat()).Gc.minor_words +. (Gc.stat()).Gc.major_words -.
                             (Gc.stat()).Gc.promoted_words )
  in
  (
      print_string(" Total memory allocated=" ^ string_of_float(lTotalMemory) ^ " bytes\n");
  )
)

let gcBench kStretchTreeDepth =

  let rec expt m n =
    if n = 0 then 1 else m * (expt m (n - 1)) in
  let treeSize i =
    (expt 2 (i + 1)) - 1 in
  let numIters i =
    (2 * (treeSize kStretchTreeDepth)) / (treeSize i) in

  let kLongLivedTreeDepth = kStretchTreeDepth - 2 in
  let kArraySize          = 4 * (treeSize kLongLivedTreeDepth) in
  let kMinTreeDepth       = 4 in
  let kMaxTreeDepth       = kLongLivedTreeDepth in

  let rec populate iDepth (Node {left = lr ; right = rr; i ; j}) =
    if iDepth <= 0 then false
    else let iDepth = iDepth - 1
         in
         (
              lr := make_empty_node();
              rr := make_empty_node();
			  (populate iDepth (!lr));
			  (populate iDepth (!rr))
		 )  in

  let rec makeTree iDepth =
     if iDepth <= 0
         then make_empty_node()
		 else make_node ( (makeTree (iDepth - 1)) ,
                          (makeTree (iDepth - 1)) )  in

  let timeConstruction depth =
    let iNumIters = numIters depth
    in (
	     print_string("Creating " ^ string_of_int(iNumIters) ^
					  " trees of depth " ^string_of_int(depth) ^ "\n");

	     let tStart = Sys.time() in
           let rec loop i =
               if i < iNumIters
                  then (  ( populate depth (make_empty_node()) ) ; loop (i + 1) )
			      else ()
		   in ( loop 0 );
           (
             let tElapsed = (Sys.time() -. tStart) in
		     print_string(  "\tTop down construction took "
			 	       	    ^ string_of_float( tElapsed *. 1000. )
					        ^ "msecs\n" )
           );
         let tStart = Sys.time() in
           let rec loop i =
               if i < iNumIters
                  then ( (makeTree depth) ; loop (i + 1) )
           in ( loop 0 );
           (
             let tElapsed = (Sys.time() -. tStart) in
			 print_string(  "\tBottom up construction took "
			     		    ^ string_of_float( tElapsed *. 1000. )
				            ^ "msecs\n" )
           )
	    ) in
  let main () = (
    print_string("Garbage Collector Test\n");
    print_string(" Stretching memory with a binary tree of depth " ^
			     string_of_int(kStretchTreeDepth) ^"\n" );
    printDiagnostics();
    let tStart = Sys.time()
    in (
          (makeTree kStretchTreeDepth);
          print_string(" Creating a long-lived binary tree of depth " ^
          string_of_int(kLongLivedTreeDepth) ^"\n");
          let longLivedTree = (make_empty_node())
          in (
              populate kLongLivedTreeDepth longLivedTree

             );
          print_string(" Creating a long-lived array of " ^
          string_of_int(kArraySize) ^ " floats\n");

          let arr = Array.make kArraySize 0.0
          in (
               print_float (Array.get arr 0); print_string("\n");
               for i = 0 to kArraySize/2 - 1 do ( Array.set arr i (1.0 /. float_of_int(i)) ) done
             );

          printDiagnostics();

          for i= kMinTreeDepth/2 to kMaxTreeDepth/2 do timeConstruction (i * 2) done;

          if longLivedTree = Empty || (Array.get arr 1000) <> 1.0 /. 1000. then
              print_string("Failed\n")
          else ();
          let tElapsed = Sys.time() -. tStart
          in (
               printDiagnostics();
               print_string( "Completed in " ^
                             string_of_float( tElapsed *. 1000.) ^
                             "ms.\n")

             )
       )
  ) in ( main () )


let () = gcBench 18






(**)

File: Scala/GCBench/src/main/scala/example/GCBench.scala
import java.io._
import scala.collection.mutable.ListBuffer
import org.rogach.scallop._

object GCBench {

    class Node(var left: Node, var right: Node) {
        def this() {
            this(null, null)
        }
    }

    val kStretchTreeDepth: Int = 18
    val kLongLivedTreeDepth: Int  = 16
    val kArraySize: Int  = 500000
    val kMinTreeDepth: Int = 4
    val kMaxTreeDepth: Int = 16
    
    // Nodes used by a tree of a given size
    def TreeSize(i : Int) : Int = {
        return ((1 << (i + 1)) - 1)
    }

    // Number of iterations to use for a given tree depth
    def NumIters(i : Int) : Int = {
        return 2 * TreeSize(kStretchTreeDepth) / TreeSize(i)
    }

    // Build tree top down, assigning to older objects.
    def Populate(iDepth : Int, thisNode : Node) {
    // var temp : Int = iDepth
        if (iDepth <= 0){
            return
        } else {
        //	temp -= 1
            thisNode.left = new Node()
            thisNode.right = new Node()
            Populate(iDepth - 1 , thisNode.left)
            Populate(iDepth - 1, thisNode.right)
        }
    }

    // Build tree bottom-up
    def MakeTree(iDepth : Int) : Node = {
        if(iDepth <= 0){
            return new Node()
        } else {
            return new Node(MakeTree(iDepth-1),MakeTree(iDepth-1))
        }
    }

    def PrintDiagnostics() {
        val lFreeMemory = Runtime.getRuntime().freeMemory()
        val lTotalMemory = Runtime.getRuntime().totalMemory()

        println(" Total memory available="
                + lTotalMemory + " bytes")
        println("  Free memory=" + lFreeMemory + " bytes")
    }

    def TimeConstruction(depth : Int){
        val root : Node = null
        var tStart: Long = 0l
        var tFinish: Long = 0l
        val iNumIters : Int = NumIters(depth)
        var tempTree : Node = null
        //println("Creating " + iNumIters +" trees of depth " + depth)
        tStart = System.currentTimeMillis()
        for( i <- 0 until iNumIters - 1){
            tempTree = new Node()
            Populate(depth, tempTree)
            tempTree = null;
        }
        tFinish = System.currentTimeMillis()
        //println("\tTop down construction took "+ (tFinish - tStart) + "msecs");
        tStart = System.currentTimeMillis()
        for( i <- 0 until iNumIters - 1){
            tempTree = MakeTree(depth)
            tempTree = null
        }
        tFinish = System.currentTimeMillis()
        //println("\tBottom up construction took "+ (tFinish - tStart) + "msecs");

    }

    def simpleloop(num: Int): Int = {
        var accum = 0
        for (i <- 1 until num) {
            accum = i
        }
        accum
    }

    def fibonacci(num: Int): Int = {
        if(num < 2){
            return 1
        } else {
            return fibonacci(num-2) + fibonacci(num-1)
        }
    }

    def start_gc_thread(num_threads : Int, tree_depth: Int, iterations: Int, gcsleep: Int, debug: Boolean) {
        for (i <- 1 until num_threads+1) {
            if (debug) println(s"starting gc thread ${i}")

            val t = new Thread(new Runnable {
                def run() {
                    var listofTimeStamps = gc_func(tree_depth, i, iterations, gcsleep, debug)
                    for (timeStamp <- listofTimeStamps.toList){
                        println(timeStamp)
                    }
                }
            })
            t.setDaemon(false)
            t.start()
        }
    }

    def gc_func(tree_depth: Int, id: Int, iterations: Int, gcsleep: Int, debug: Boolean): ListBuffer[String] = {
        var listOfTimeStamps = new ListBuffer[String]()
        for (i <- 1 until iterations+1) {
            if (debug) println(s"gc iter ${i}")

            val start = System.currentTimeMillis()
            val runtime1 = Runtime.getRuntime
            listOfTimeStamps += "gc:start:"+ id +":" + i+ ":" +start + ":" + (runtime1.totalMemory - runtime1.freeMemory)
            val root: Node = null
            var longLivedTree: Node = null
            var tempTree: Node = null
            var tStart: Long = 0l
            var tFinish: Long = 0l
            var tElapsed: Long = 0l

            //println("Garbage Collector Test")
            //println(" Stretching memory with a binary tree of depth "+ kStretchTreeDepth)
            //PrintDiagnostics()
            tStart = System.currentTimeMillis()

            // Stretch the memory space quickly
            tempTree = MakeTree(tree_depth) //kStretchTreeDepth)
            tempTree = null

            // Create a long lived object
            //println(" Creating a long-lived binary tree of depth " +kLongLivedTreeDepth)
            longLivedTree = new Node()
            Populate(tree_depth, longLivedTree)

            // Create long-lived array, filling half of it
            //println(" Creating a long-lived array of "+ kArraySize + " doubles")
            var array = new Array[Double](kArraySize)
            for ( i <- 0 until kArraySize/2) {
                array(i) = 1.0/i
            }
            //PrintDiagnostics()

            var d : Int = kMinTreeDepth
            while(d <= tree_depth){
                TimeConstruction(d)
                d += 2
            }

            if (longLivedTree == null || array(1000) != 1.0 / 1000) println("Failed")

            val runtime2 = Runtime.getRuntime
            tFinish = System.currentTimeMillis()
            tElapsed = tFinish-tStart
            //PrintDiagnostics()
            //println("Completed in " + tElapsed + "ms.")
            val stop = System.currentTimeMillis()
            listOfTimeStamps += "gc:stop:"+ id +":" + i + ":" + stop + ":" + (runtime2.totalMemory - runtime2.freeMemory)

            Thread.sleep(gcsleep * 1000)
        }
        listOfTimeStamps
    }

    def start_comp_threads(num_threads : Int, depth: Int, iterations: Int, comp_sleep: Int, debug: Boolean){
        for (i <- 1 until num_threads+1) {
            if (debug) println(s"starting comp thread ${i}")

            val t = new Thread(new Runnable {
                def run() {
                    var listofTimeStamps = comp_func(depth, i, iterations, comp_sleep, debug)
                    for (timeStamp <- listofTimeStamps.toList){
                        println(timeStamp)
                    }
                }
            })
            t.setDaemon(false)
            t.start()
        }
    }

    def comp_func(depth: Int, id: Int, iterations: Int, comp_sleep: Int, debug: Boolean): ListBuffer[String] = {
        var listOfTimeStamps = new ListBuffer[String]()
        for (i <- 1 until iterations+1) {
            if (debug) println(s"comp iter ${i}")

            val tStart = System.currentTimeMillis()
            val runtime3 = Runtime.getRuntime
            listOfTimeStamps += ("compute:start:"+ id +":" + i + ":" +tStart  + ":" + (runtime3.totalMemory - runtime3.freeMemory))
            //fibonacci(depth)
            simpleloop(depth)
            val tStop = System.currentTimeMillis()
            val runtime4 = Runtime.getRuntime
            listOfTimeStamps += ("compute:stop:"+ id +":" + i + ":" +tStop  + ":" + (runtime4.totalMemory - runtime4.freeMemory))
            Thread.sleep(comp_sleep * 1000)
        }
        listOfTimeStamps
    }

    class Conf(arguments: Seq[String]) extends ScallopConf(arguments) {
        val computeThreads = opt[Int](default = Some(1), short = 't')
        val computeDepth = opt[Int](default = Some(40), short = 'd')
        val iterations = opt[Int](default = Some(1100), short = 'i')
        val computeSleep = opt[Int](default = Some(1), short = 's')
        val gcThreads = opt[Int](default = Some(1), short = 'g')
        val gcSleep = opt[Int](default = Some(1), short = 'S')
        val gcDelay = opt[Int](default = Some(60), short = 'J')
        val treeDepth = opt[Int](default = Some(15), short = 'G')

        var debug = opt[Boolean](default = Some(false), short='D')
        val help = opt[Boolean]()

        version("Scala GCBench 0.1.0")
        banner("""Usage: GCBench [OPTION]
                |Options:
                |""".stripMargin)
        footer("\nhttps://ubmltongroup.github.io/")

        verify()
    }

    def main(args: Array[String]) {
        val conf = new Conf(args)
        if (conf.help()) {
            conf.printHelp();
        }
        else {
            if (conf.debug()) println("Start compute threads")
            start_comp_threads(conf.computeThreads(), conf.computeDepth(), conf.iterations(),
                               conf.computeSleep(), conf.debug())

            Thread.sleep(conf.gcDelay() * 1000)

            if (conf.debug()) println("Start GC threads")
            start_gc_thread(conf.gcThreads(), conf.treeDepth(), conf.iterations(),
                            conf.gcSleep(), conf.debug())
        }
    }
}
